src/uct/ib/ud/verbs/ud_verbs.c

```diff
@@ -373,23 +383,33 @@ uct_ud_verbs_iface_poll_tx(uct_ud_verbs_iface_t *iface, int is_async)
         return 0;
     }
 
-    if (ucs_unlikely(wc.status != IBV_WC_SUCCESS)) {
-        ucs_fatal("Send completion (wr_id=0x%0X with error: %s ",
-                  (unsigned)wc.wr_id, ibv_wc_status_str(wc.status));
-        return 0;
-    }
+    for (i = 0; i < ret; i++) {
+        if (ucs_unlikely(wc[i].status != IBV_WC_SUCCESS)) {
+            ucs_fatal("Send completion (wr_id=0x%0X with error: %s ",
+                      (unsigned)wc[i].wr_id, ibv_wc_status_str(wc[i].status));
+            continue;
+        }
 
-    UCS_STATS_UPDATE_COUNTER(iface->super.super.stats,
-                             UCT_IB_IFACE_STAT_TX_COMPLETION, 1);
+        UCS_STATS_UPDATE_COUNTER(iface->super.super.stats,
+                                 UCT_IB_IFACE_STAT_TX_COMPLETION, 1);
+
+        if (uct_ib_iface_device(&iface->super.super)->ordered_send_comp) {
```

Comment:
to avoid extra branches in progress can we have two different progress function for efa and IB UD ifaces?

Reply:
after checking, difficult since first function pointer is top level ud verbs iface progress